cf-file-scanner/src/main/java/org/broadinstitute/ddp/cf/FileScanner.java

```diff
@@ -0,0 +1,280 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.cloud.functions.BackgroundFunction;
+import com.google.cloud.functions.Context;
+import com.google.cloud.pubsub.v1.Publisher;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.ProjectTopicName;
+import com.google.pubsub.v1.PubsubMessage;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+
+public class FileScanner implements BackgroundFunction<FileScanner.Message> {
+
+    private static final Logger logger = Logger.getLogger(FileScanner.class.getName());
+
+    private static final String APP_CONF = "./application.conf";
+    private static final String ENV_CONF_FILE = "config.file";
+    private static final String CFG_REFRESH_TIME = "refreshTime";
+    private static final String CFG_REFRESH_UNIT = "refreshUnit";
+    private static final String CFG_GCP_PROJECT_ID = "gcpProjectId";
+    private static final String CFG_RESULT_TOPIC = "resultTopic";
+
+    private static final String ATTR_BUCKET_ID = "bucketId";
+    private static final String ATTR_OBJECT_ID = "objectId";
+    private static final String ATTR_SCAN_RESULT = "scanResult";
+
+    private static final String FRESHCLAM_CONF = "./freshclam.conf";
+    private static final String FRESHCLAM_BIN = "./clamav/freshclam";
+    private static final String CLAMSCAN_BIN = "./clamav/clamscan";
+    private static final String BIN_DIR = "./clamav";
+    private static final String DB_DIR = "/tmp/clamav-db";
+
+    private static final long refreshMillis;
+    private static final Storage storage;
+    private static final Publisher publisher;
+    private static Instant dbUpdatedAt = null;
+
+    // This is ran once on cold start.
+    static {
+        Config cfg;
+        if (System.getenv(ENV_CONF_FILE) != null) {
+            cfg = ConfigFactory.load();
+        } else {
+            cfg = ConfigFactory.parseFile(new File(APP_CONF));
+        }
+
+        TimeUnit unit = TimeUnit.valueOf(cfg.getString(CFG_REFRESH_UNIT));
+        if (unit == TimeUnit.MICROSECONDS || unit == TimeUnit.NANOSECONDS) {
+            throw new RuntimeException("Refresh time unit is too granular: " + unit);
+        }
+        long refreshTime = cfg.getLong(CFG_REFRESH_TIME);
+        refreshMillis = unit.toMillis(refreshTime);
+
+        String gcpProjectId = cfg.getString(CFG_GCP_PROJECT_ID);
+        String resultTopic = cfg.getString(CFG_RESULT_TOPIC);
+        try {
+            storage = StorageOptions.newBuilder()
+                    .setCredentials(GoogleCredentials.getApplicationDefault())
+                    .setProjectId(gcpProjectId)
```

Comment:
I'm not sure if setting the project here is necessary - I'd double-check if the storage client requires it.